test/common/http/codec_client_test.cc

```diff
@@ -167,5 +172,146 @@ TEST_F(CodecClientTest, WatermarkPassthrough) {
   connection_cb_->onBelowWriteBufferLowWatermark();
 }
 
+// Test the codec getting input from a real TCP connection.
+class CodecNetworkTest : public testing::Test {
+public:
+  CodecNetworkTest() {
+    dispatcher_.reset(new Event::DispatcherImpl);
+    upstream_listener_ =
+        dispatcher_->createListener(connection_handler_, socket_, listener_callbacks_, stats_store_,
+                                    Network::ListenerOptions::listenerOptionsWithBindToPort());
+    Network::ClientConnectionPtr client_connection =
+        dispatcher_->createClientConnection(socket_.localAddress(), source_address_);
+    client_connection_ = client_connection.get();
+    codec_ = new Http::MockClientConnection();
+    client_.reset(new CodecClientForTest(std::move(client_connection), codec_, nullptr, host_));
+    EXPECT_CALL(listener_callbacks_, onNewConnection_(_))
+        .WillOnce(Invoke([&](Network::ConnectionPtr& conn) -> void {
+          upstream_connection_ = std::move(conn);
+          upstream_connection_->addConnectionCallbacks(upstream_callbacks_);
+          dispatcher_->exit();
+        }));
+    dispatcher_->run(Event::Dispatcher::RunType::Block);
+  }
+
+  void createNewStream() {
+    Http::StreamDecoder* inner_decoder;
+    EXPECT_CALL(*codec_, newStream(_))
+        .WillOnce(Invoke([&](Http::StreamDecoder& decoder) -> Http::StreamEncoder& {
+          inner_decoder = &decoder;
+          return inner_encoder_;
+        }));
+
+    client_->newStream(outer_decoder_);
+  }
+
+  void close() {
+    client_->close();
+    EXPECT_CALL(upstream_callbacks_, onEvent(Network::ConnectionEvent::RemoteClose))
+        .WillOnce(InvokeWithoutArgs([&]() -> void { dispatcher_->exit(); }));
+    dispatcher_->run(Event::Dispatcher::RunType::Block);
+  }
+
+protected:
+  Network::MockListenerCallbacks listener_callbacks_;
+  Network::MockConnectionHandler connection_handler_;
+  Network::Address::InstanceConstSharedPtr source_address_;
+  Stats::IsolatedStoreImpl stats_store_;
+  Network::ListenerPtr upstream_listener_;
+  Event::DispatcherPtr dispatcher_;
+  Network::TcpListenSocket socket_{Network::Test::getAnyAddress(Network::Address::IpVersion::v6),
```

Comment:
Is this IPv6 only? What about IPv4 only setups? Should the test be paramaterized like we do elsewhere to handle IPv4/IPv6 setup variations?